### ISSUE TITLE
Can BemRosetta show the curve on the line paralleled to diagonal of QTF matrix?  

### ISSUE THREAD

**Issue by getChaos22** [BEMRosetta/BEMRosetta#38]
Hello @izabala123,

BemRosetta can show the curves on the diagonal line, horizontal line and vertical line of a QTF matrix. But often we concern more about the values on the line paralleled to diagonal or back-diagonal, because they represent same difference-frequency or sum-frequency. So, is there a way to displace curve on the line paralleled to diagonal in BemRosetta?

Thank you!

Chao Li

**Comment by izabala123** [comment]
Hi Chao

It looks interesting. I will implement it.

Best regards
Iñaki

**Comment by izabala123** [comment]
Hi Chao

Thinking about your request, I wanted to ask you how you would prefer the data to be displayed. 

In the current implementation, the vertical axis shows the QTF value, and the horizontal axis shows, for the horizontal line the frequency on the horizontal axis, for the vertical line the frequency on the vertical axis, for the diagonal it doesn't matter, the horizontal and vertical show the same thing. However, for what you propose, what would you prefer to be shown on the horizontal axis?

Best regards
Iñaki

**Comment by getChaos22** [comment]
Hi @izabala123,

I think it's both OK to show horizontal frequency or vertical frequency on the horizontal axis, because we can calculate another frequency basing on the difference-frequency or sum-frequency equation. Some materials (like Nemoh's manual) show vertical frequency on the horizontal axis, but maybe it's more intuitive to show horizontal frequency because the start frequency is constant? Anyway, I think it's both acceptable to show horizontal frequency or vertical frequency on the horizontal axis.

Best regards
Chao Li

**Comment by izabala123** [comment]
Hi Chao

By including this feature I have noticed a problem. As you can see, it is possible that the difference between frequencies is not constant. In this case, when drawing the line that passes through values parallel to the main diagonal, this line is not straight. Is it better to do it this way, or to show a straight line, whose values are a bilinear interpolation?

![image](https://github.com/BEMRosetta/BEMRosetta/assets/38589221/7336f3ca-33fd-4796-a58b-a668141427f7)

Best regards
Iñaki

**Comment by getChaos22** [comment]
Hi Iñaki

Yeah, you're right. When the frequency interval is not constant, the line that represents difference between two frequencies with same identifier difference (like w1-w2, w2-w3, w3-w4,.... or  w1-w3, w2-w4, w3-w5,....  ) is not straight and parallel to the main diagonal. The line only parallel to the main diagonal when the frequency interval is constant since the slope is 1. 

I think it is more reasonable to show data on straight lines parallel to the main diagonal rather than data with same difference of frequency identifier. Because the first way is closer to the definition of difference-frequency. So I think it is better to show a straight line, even though the values may be obtained by interpolation.

 BTW, is it a common situation that the frequency interval is not constant? Because in my experience, usually we can only set constant frequency interval in some hydrodynamics softwares (like Nemoh, WAMIT and AQWA).

Best Regards

Chao Li

**Comment by izabala123** [comment]
Hi Chao

I have observed that some researchers prefer to set the values by period and not by frequency. In addition, many options are available in Wamit, with the IPERIN - IPEROUT labels, including variable interval values of (from Wamit manual):

![image](https://github.com/BEMRosetta/BEMRosetta/assets/38589221/7e647c23-40c0-43c5-9ca7-b4c75b93c788)

I will thus consider the interpolated values taken by the straight line.

Best regards
Iñaki

**Comment by izabala123** [comment]
Dear Chao

BEMRosetta is updated. Please test and give me your feedback.

![image](https://github.com/BEMRosetta/BEMRosetta/assets/38589221/fd76af92-16dc-4017-b328-b69e1ee5fd3e)

Best regards
Iñaki

**Comment by getChaos22** [comment]
Hi Iñaki

Thanks so much for the update of BEMRosetta! 

I tested the new version of BemRosseta. There are some questions I came across and would like to provide you with feedback.

The first question is that the QTF curve exhibits oscillations at high frequency range both at the diagonal and off-diagonal. This   
phenomenon does not happen in the previous version at the diagonal. 
![image](https://github.com/BEMRosetta/BEMRosetta/assets/120010727/28e3ac34-9ec1-44ca-ac02-705ed3d530aa)

The second question is that when the QTF matrix is stored in a symmetrical form (like the result files I provided in the attachment), the QTF curve shows nothing both at the diagonal and off-diagonal. 
![image](https://github.com/BEMRosetta/BEMRosetta/assets/120010727/81e10243-5885-4d9d-b627-d38953c15d64)

The last question also relates to the QTF matrix stored in symmetrical form . The QTF values in the left table show a distribution that is neither upper-triangle nor lower-triangle. It seems like many elements have been omitted or arranged in a different way.
![image](https://github.com/BEMRosetta/BEMRosetta/assets/120010727/4085c139-c17d-437e-a191-f3ac5596288e)

Best regards

Chao Li

**Comment by izabala123** [comment]
Hi Chao

Sorry for the delay. The problems you indicate are solved, so I hope to upload a corrected version in the next few days. I will also include an explanation of the cause of the problems.

Best regards
Iñaki

**Comment by izabala123** [comment]
Hi Chao

An updated version is available in "January" binaries. The fixes have been:
- From now on, if a file has incomplete QTFs, they are filled with the values on the other side of the diagonal. This is based on the assumption that the body is still. This approach is more general, as not all solvers accept that only half a matrix is available. In addition, it prevents the program from going crazy, as empty spaces, which are null values, were processed as if they were real values, distorting the internal calculations.
- The interpolated values are taken only for X-values which coincide with the true frequencies, thus avoiding oscillations. These were caused by bilinear interpolation with very different values. For example:
![image](https://github.com/BEMRosetta/BEMRosetta/assets/38589221/3d8dfa33-72ee-4073-980c-794181bc9d1a)
If we zoom at the right tail, we see that the values taken have much higher values around them. A bilinear interpolation, that in fact is not linear, will create a kind of saw-teeth because of the closeness of the higher values. Now, taking the values just in the x of the real frequencies, this is avoided.
![image](https://github.com/BEMRosetta/BEMRosetta/assets/38589221/083390da-77c6-47ea-86e8-fec00b8304e4)
Along with this, the entire QTF display code has been modified, as in its previous state it was full of patches. All this has not affected the internal processing of QTF, only its display.

I hope that all this will solve the problems encountered. Thank you very much for your patience.

Best regards
Iñaki

**Comment by getChaos22** [comment]
Hi @izabala123,

Thank you so much for putting a lot of efforts on this requirement. I have downloaded the new version of BEMRosetta and tested it. All the problems I came across at the last version had been fixed perfectly. So now I can use BEMRosetta to conduct deeper comparison and research between different BEM solvers. 
 
We appreciate again for the new feature of BEMRosetta. It has brought us great convenience.

Best regards
Chao Li
